### PR TITLE
Enable support DllImport a native assembly whose name contains '.'

### LINKED ISSUE #17150: Not possible to DllImport a native assembly with a "." in its file name
Since our tooling doesn't support customers to compile for each OS, it currently isn't possible to name a native assembly using namespace names, like "Microsoft.Foo.Bar", and use the assembly in a `DllImport` successfully on both Windows and non-Windows.

## Repro:
Create a native assembly named `Test.Foo`.  On Windows it is named `Test.Foo.dll`.  On Linux it is named `Test.Foo.so`.  On macOS it is named `Test.Foo.dylib`.

It is not possible to create a C# `DllImport` into this native assembly from a single managed assembly.

On Windows, if you use:

```C#
        internal const string NativePath = "Test.Foo";

        [DllImport(NativePath)]
        public static extern bool Bar();
```

You get an error:

`System.DllNotFoundException : Unable to load DLL 'Test.Foo': The specified module could not be found. (Exception from HRESULT: 0x8007007E)`

So you need to use `Test.Foo.dll` in order for Windows to load it.

However, doing this causes it to fail on non-Windows.  On non-Windows if you use:

```C#
        internal const string NativePath = "Test.Foo.dll";

        [DllImport(NativePath)]
        public static extern bool Bar();
```

You get the error:

`System.DllNotFoundException: Unable to load shared library 'Test.Foo.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libTest.Foo.dll: cannot open shared object file: No such file or directory`

Using `strace`, it shows that we are not stripping the `.dll` off of the name during probing:

```
open("/mnt/shared/TestML/runtimes/linux-x64/native/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/eerhardt/dotnet/shared/Microsoft.NETCore.App/2.1.0-preview3-26322-05/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/mnt/shared/TestML/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 82
fstat(82, {st_mode=S_IFREG|0644, st_size=95952, ...}) = 0
mmap(NULL, 95952, PROT_READ, MAP_PRIVATE, 82, 0) = 0x7ff76fd9f000
close(82)                               = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/x86_64-linux-gnu/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/Test.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
munmap(0x7ff76fd9f000, 95952)           = 0
open("/mnt/shared/TestML/runtimes/linux-x64/native/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/eerhardt/dotnet/shared/Microsoft.NETCore.App/2.1.0-preview3-26322-05/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/mnt/shared/TestML/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 82
fstat(82, {st_mode=S_IFREG|0644, st_size=95952, ...}) = 0
mmap(NULL, 95952, PROT_READ, MAP_PRIVATE, 82, 0) = 0x7ff76fd9f000
close(82)                               = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/x86_64-linux-gnu/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/libTest.Foo.dll.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
munmap(0x7ff76fd9f000, 95952)           = 0
open("/mnt/shared/TestML/runtimes/linux-x64/native/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/eerhardt/dotnet/shared/Microsoft.NETCore.App/2.1.0-preview3-26322-05/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/mnt/shared/TestML/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 82
fstat(82, {st_mode=S_IFREG|0644, st_size=95952, ...}) = 0
mmap(NULL, 95952, PROT_READ, MAP_PRIVATE, 82, 0) = 0x7ff76fd9f000
close(82)                               = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/x86_64-linux-gnu/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/Test.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
munmap(0x7ff76fd9f000, 95952)           = 0
open("/mnt/shared/TestML/runtimes/linux-x64/native/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/home/eerhardt/dotnet/shared/Microsoft.NETCore.App/2.1.0-preview3-26322-05/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/mnt/shared/TestML/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 82
fstat(82, {st_mode=S_IFREG|0644, st_size=95952, ...}) = 0
mmap(NULL, 95952, PROT_READ, MAP_PRIVATE, 82, 0) = 0x7ff76fd9f000
close(82)                               = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/x86_64-linux-gnu/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/libTest.Foo.dll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
munmap(0x7ff76fd9f000, 95952)           = 0
```

## Possible solution

We could add another variation on non-Windows that stripped the `.dll` off of the library name, and then append `.so` or `.dylib` to it.  This would allow the name `Test.Foo.dll` to work on all currently supported platforms.

/cc @jkotas @stephentoub

### PR DESCRIPTION
**Pull request by luqunl**
Currently, In Windows, if the name your native dll specified in DllImport contains '.', such as Test.Foo, DllImport will only search for file Test.Foo and it won't search for file Test.Foo.dll

The fix is to append ".dll" if the native dll name specified in DllImport doesn't contain ".dll" .

Fix #17150

### CODE REVIEW

**Comment by jkotas** [comment]
Test?

**Comment by danmosemsft** [comment]
Ideally a test in Corefx so we can ensure it is fixed in Mono, CoreRT.

**Comment by jkotas** [comment]
CoreCLR has better setup for interop tests. And Mono and CoreRT can run CoreCLR tests too.

**Comment by mikedn** [comment]
> The fix is to append ".dll" if the native dll name specified in DllImport doesn't contain ".dll" .

So what happens if the specified name is "foo.exe"? Will it search for "foo.exe.dll"?

**Comment by kasper3** [comment]
I totally missed your PR. This approach of using `DetermineLibNameVariations` for both Windows and Unix is most probably neater than what I was experimenting with https://github.com/kasper3/coreclr/commit/edb45878f6915bf7443d74f7da2d17dd1dae5ac9, except for missing `.exe` check.
Also, does it account for casing `.dll`, `.DLL` etc.?

**Comment by jkotas** [comment]
@luqunl Could you please rebase away the commits not related to your change?

**Comment by luqunl** [comment]
Let me try to rebase my changes

**Comment by luqunl** [comment]
> Could you please rebase away the commits not related to your change?
Fixed

> Test?
Fixed

> So what happens if the specified name is "foo.exe"? Will it search for "foo.exe.dll"?
Thanks for catching this. Fixed

> Also, does it account for casing .dll, .DLL etc.?
Thanks for catching this. Fixed

File: src/vm/dllimport.cpp [hunk 1]
```diff
@@ -6042,20 +6052,26 @@ HMODULE NDirect::LoadFromNativeDllSearchDirectories(AppDomain* pDomain, LPCWSTR
...
         SString::CIterator it = libName.Begin();
-        if (libName.FindASCII(it, PAL_SHLIB_SUFFIX))
+        if (libName.FindASCII(it, PLATFORM_SHARED_LIB_SUFFIX_A))
         {
-            it += strlen(PAL_SHLIB_SUFFIX);
+            it += (int)strlen(PLATFORM_SHARED_LIB_SUFFIX_A);
             containsSuffix = it == libName.End() || *it == (WCHAR)'.';
         }
+#else
+        if (libName.EndsWithCaseInsensitive(PLATFORM_SHARED_LIB_SUFFIX_W) || libName.EndsWithCaseInsensitive(WINDOWS_SHARED_EXE_SUFFIX_W))
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: I do not understand why we are special casing `.exe` suffix here. `.exe` suffix is non-standard suffix for shared libraries. If I want to, I can use `.dat` suffix for my .dlls and LoadLibrary will load them too.

File: tests/src/Interop/DllImportAttribute/Simple/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required (VERSION 2.6)
+project (DllFileProbe)
+set(SOURCES DllFileProbe.cpp )
+
+# add the executable
+add_library (DllFileProbe SHARED ${SOURCES})
+target_link_libraries(DllFileProbe ${LINK_LIBRARIES_ADDITIONAL}) 
+
+#get_cmake_property(_variableNames VARIABLES)
```
Comment (jkotas) [inline tests/src/Interop/DllImportAttribute/Simple/CMakeLists.txt]: Do we need these commented out blocks?

Comment (luqunl) [reply tests/src/Interop/DllImportAttribute/Simple/CMakeLists.txt]: I will remove these commented lines..just copy & paste..

File: src/vm/dllimport.cpp (hunk 1 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: The only reason is to avoid to load/search .exe.dll file. if it is possible that someone name his file as a.exe.dll, then we can remove this special case.

Comment (jkotas) [reply src/vm/dllimport.cpp]: I am more worried about that cases that used to work fine before this change, but do not work anymore after this change. Consider:

```csharp
[DllImport("mydll.dat")]
extern static int sum(int a, int b);

static void Main() {
    Console.WriteLine(sum(1,2).ToString());
}
```

And `mydll.dat` that is renamed `mydll.dll` to `mydll.dat`. It works fine before this change (`mydll.dat` is passed to LoadLibrary), but it does not work after this change (`mydll.dat.dll` is passed to LoadLibrary).

Comment (luqunl) [reply src/vm/dllimport.cpp]: Okay. 
`
            static const char* const SuffixFirst[] =
             {
                 "%.0s%s%s", // name+suffix
#ifdef FEATURE_PAL
                 "%s%s%s",   // prefix+name+suffix
#endif
                 "%.0s%s",   // name
#ifdef FEATURE_PAL
                 "%s%s%.0s"  // prefix+name
#endif
             };
'

File: tests/src/Interop/DllImportAttribute/Simple/CMakeLists.txt (hunk 2 above)
Comment (luqunl) [reply tests/src/Interop/DllImportAttribute/Simple/CMakeLists.txt]: Removed these comments

File: src/vm/dllimport.cpp (hunk 1 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: I can check // name first, then //name + suffix to be more comp with current behavior.

File: src/vm/dllimport.cpp [hunk 3]
```diff
@@ -6078,9 +6096,13 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
             static const char* const SuffixFirst[] =
             {
                 "%.0s%s%s", // name+suffix
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: For name without any dots, this is going to probe twice now:

- First time, we append the `.dll`
- Second time, we pass in name without `.dll` suffix and LoadLibrary appends it internally.

It would be nice to avoid this duplication (e.g. just let LoadLibrary take care of appending .dll to the simple names).

File: src/vm/dllimport.cpp [hunk 4]
```diff
@@ -6078,9 +6096,13 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
...
+#ifdef FEATURE_PAL
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: Similar for names that end with a `.`. We should not do any probing for them on Windows - to maintain the behavior documented for them on MSDN.

File: src/vm/dllimport.cpp [hunk 5]
```diff
@@ -5890,6 +5890,16 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
 #define TOLOWER(a) (((a) >= W('A') && (a) <= W('Z')) ? (W('a') + (a - W('A'))) : (a))
 #define TOHEX(a)   ((a)>=10 ? W('a')+(a)-10 : W('0')+(a))
 
+#ifdef FEATURE_PAL
+#define PLATFORM_SHARED_LIB_SUFFIX_A PAL_SHLIB_SUFFIX
+#define PLATFORM_SHARED_LIB_PREFIX_A PAL_SHLIB_PREFIX
+#else // !FEATURE_PAL
+#define PLATFORM_SHARED_LIB_SUFFIX_A ".dll"
+#define PLATFORM_SHARED_LIB_PREFIX_A ""
+#define PLATFORM_SHARED_LIB_SUFFIX_W W(".dll")
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: Nit: I do not think we need to have a defines for these since they are used just once.

File: src/vm/dllimport.cpp (hunk 3 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Fixed

File: src/vm/dllimport.cpp [hunk 6]
```diff
@@ -6118,6 +6104,36 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
         *numberOfVariations = COUNTOF(NameOnly);
     }
 }
+#else
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: Nit: It would be nice to add `// FEATURE_PAL` comment here.

File: src/vm/dllimport.cpp [hunk 7]
```diff
@@ -6118,6 +6104,36 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
...
+        static const char* const SuffixLast[] =
+        {
+            "%.0s%s",   // name
+            "%.0s%s%s", // name+suffix
+        };
+        *libNameVariations = SuffixLast;
+        *numberOfVariations = COUNTOF(SuffixLast);
+    }
+}
+#endif
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: Same here.

File: src/vm/dllimport.cpp [hunk 8]
```diff
@@ -6118,6 +6104,36 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
...
 }
+#else
+static void DetermineLibNameVariations(const char* const** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    bool endWithDot = libName.EndsWith(W("."));
+    bool notContainDot = !libName.Find(libName.Begin(), W('.'));
+    bool containsSuffix = libName.EndsWithCaseInsensitive(W(".dll")) || libName.EndsWithCaseInsensitive(W(".exe"));
+    if (endWithDot || notContainDot || containsSuffix || !libNameIsRelativePath)
+    {
+        // file name end with '.'
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: This comment is not very useful. It is a literal rewrite of the above condition in plain English.

Instead, it would be useful to have comment that describes the reasons for this logic.

File: src/vm/dllimport.cpp (hunk 6 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Updated

File: src/vm/dllimport.cpp (hunk 7 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Updated

File: src/vm/dllimport.cpp (hunk 8 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Added more comments

**Comment by jkotas** [comment]
Build errors:

```
src\vm\dllimport.cpp(6110): error C2220: warning treated as error - no 'object' file generated 
src\vm\dllimport.cpp(6110): warning C4800: 'BOOL': forcing value to bool 'true' or 'false' (performance warning)
```

**Comment by luqunl** [comment]
Fixed build break for BOOL to bool conversation.

File: src/vm/dllimport.cpp [hunk 9]
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
...
-#endif
+#else // FEATURE_PAL
+static void DetermineLibNameVariations(const char* const** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    bool endWithDot = libName.EndsWith(W(".")) == TRUE;
+    bool notContainDot = !libName.Find(libName.Begin(), W('.'));
+    bool containsSuffix = libName.EndsWithCaseInsensitive(W(".dll")) || libName.EndsWithCaseInsensitive(W(".exe"));
+    
+    // The purpose of following code is to workaround LoadLibrary limitation: 
+    // LoadLibray won't append extension if filename itself contains '.'. Thus It will break the following scenario:
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: Typo:  It -> it

File: src/vm/dllimport.cpp [hunk 10]
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
         *numberOfVariations = COUNTOF(NameOnly);
     }
 }
-#endif
+#else // FEATURE_PAL
+static void DetermineLibNameVariations(const char* const** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    bool endWithDot = libName.EndsWith(W(".")) == TRUE;
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: It is a bad practice to compare `BOOL` to `TRUE`. `BOOL` is actually integer and any non-zero value can mean true. `TRUE` is defined as `1`. So comparing `BOOL` to `TRUE` assumes that the non-zero true value is `1` that is not always the case.

Use either `!!libName.EndsWith(W("."))`  or `libName.EndsWith(W(".")) != FALSE`.

**Comment by luqunl** [comment]
> tests/src/Interop/DllImportAttribute/FileExtensionProbe.cs is encoded as UTF16LE (file --mime-encoding FileExtensionProbe.cs), and github is showing it as a binary file in diff view. You can fix it by changing it to ASCII using notepad file>saveAs or bash: iconv -f utf-16le -t US_ASCII FileExtensionProbe.cs > FileExtensionProbe.cs_new then delete the old and rename _new one.

Thanks a lot for this info. I saw this issue but didn't realize that this is file encoding issue.

File: src/vm/dllimport.cpp (hunk 10 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Thanks for this. Previous I had searched the implementation for SString.EndWith and it always return TRUE or FALSE.  I will use !! instead.

**Comment by luqunl** [comment]
> Typo in loadlibrary
Fixed

File: src/vm/dllimport.cpp (hunk 9 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Fixed

**Comment by kasper3** [comment]
@jkotas, is there a chance to include this in 2.1 release?

File: src/vm/dllimport.cpp [hunk 11]
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
         *numberOfVariations = COUNTOF(NameOnly);
     }
 }
-#endif
+#else // FEATURE_PAL
+static void DetermineLibNameVariations(const char* const** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    bool endWithDot = !!libName.EndsWith(W("."));
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: Now that I am looking at this again ... caching these in bool variables makes this less efficient than it needs to be. If you put these directly into the if statement without caching them in locals like this:

```
if (!libName.Find(libName.Begin(), W('.')) ||
    libName.EndsWith(W(".")) ||
    libName.EndsWithCaseInsensitive(W(".dll")) || libName.EndsWithCaseInsensitive(W(".exe"))
{
```

It will be more efficient because of it will stop at `!libName.Find(libName.Begin(), W('.'))` for names without dot; and the remaining two statements won't be executed at all.

It also avoids the `!!` wart.

**Comment by jkotas** [comment]
> @jkotas, is there a chance to include this in 2.1 release?

@kasper3 Why do you think it is important to include this in 2.1 release?

**Comment by kasper3** [comment]
> There would need to be a good justification why this change is important to include in 2.1

In 2.0 and older versions, we are using build time conditions:

```c#
#if WINNT
[DllImport("shared.library.dll")]
#else
[DllImport("shared.library")]
#endif
```

With this change, it will be reduced to `[DllImport("shared.library")]`. The sooner this inconsistency goes away, the better.

File: src/vm/dllimport.cpp [hunk 12]
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
...
+{
+    // The purpose of following code is to workaround LoadLibrary limitation: 
+    // LoadLibrary won't append extension if filename itself contains '.'. Thus it will break the following scenario:
+    // [DllImport("A.B")] // The full name for file is "A.B.dll". This is common code pattern for cross-platform PInvoke
+    // The workaround for above scenario is to call LoadLibrary with "A.B" first, if it fails, then call LoadLibrary with "A.B.dll"
+    if (!libName.Find(libName.Begin(), W('.')) || 
+        libName.EndsWith(W(".")) || 
+        libName.EndsWithCaseInsensitive(W(".dll")) || 
+        libName.EndsWithCaseInsensitive(W(".exe")) || 
+        !libNameIsRelativePath)
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: `libNameIsRelativePath` should be first because of it is the cheapest condition to evaluate.

Comment (jkotas) [reply src/vm/dllimport.cpp]: Also, is the relative path condition really correct here? Relative path is going to be for a path like:

`XXX\YYY\library`

or

`\XXX\YYY\library`.

On Unix, we will end up trying paths like:

`libXXX\YYY\library.so`

or 

`lib\XXX\YYY\library`.

File: src/vm/dllimport.cpp [hunk 13]
```diff
@@ -6175,11 +6217,7 @@ HINSTANCE NDirect::LoadLibraryModule(NDirectMethodDesc * pMD, LoadLibErrorTracke
     for (int i = 0; hmod == NULL && i < numberOfVariations; i++)
     {
         SString currLibNameVariation;
-        currLibNameVariation.Printf(prefixSuffixCombinations[i], PAL_SHLIB_PREFIX, name, PAL_SHLIB_SUFFIX);
-#else
-    {
-        LPCWSTR currLibNameVariation = wszLibName;
-#endif
+        currLibNameVariation.Printf(prefixSuffixCombinations[i], PLATFORM_SHARED_LIB_PREFIX_A, name, PLATFORM_SHARED_LIB_SUFFIX_A);
```
Comment (jkotas) [inline src/vm/dllimport.cpp]: We have converted the name from UTF8 to Unicode above once already.

The PrintF is going to convert it again; but I do not think it is going to use UTF8 on Windows. I think this needs to use the converted name to Unicode (wszLibName); not the UTF8 name.

**Comment by jkotas** [comment]
> include this in 2.1 release

There seems to be way too many subtle issues in this change. I do not think we should be considering this for 2.1 release.

File: src/vm/dllimport.cpp (hunk 13 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Update all of these argument to WCHAR* instead of char*

File: src/vm/dllimport.cpp (hunk 12 above)
Comment (luqunl) [reply src/vm/dllimport.cpp]: Move libNameIsRelativePath  to the first one. 

 'is the relative path condition really correct here?

This is incorrect. since this is an old issue and I will create another issue to track and solve this problem

**Comment by jkotas** [comment]
LGTM

**Comment by jkotas** [comment]
master is open for 2.2 now. Merge means that we believe it is a good change for 2.2.

It would be up to @luqunl to present it to the shiproom to consider it for 2.1. My guess is that it would not meet the bar.

**Comment by sdmaclea** [comment]
This file introduced corrupt line endings https://github.com/dotnet/coreclr/pull/17505/files#diff-23d6cee28d9638de3cc99b9da9fbb3b9
